### PR TITLE
ISSUE 87: vf2 algorithm refactor

### PR DESCRIPTION
**Pull request by sracha4355**
I've restructured the code by splitting up the responsibilities of the algorithm into multiple classes. There are three major components. 
1. Generating candidate vertex pairs based the state of the search space for this algorithm
2. Updating the sets and mappings that help the algorithm narrow or broaden the search space
3. Creating a mapping where we map all vertex ids in a graph to the range 1...N where N is the number of vertices in a graph
4. Checking if a potential candidate vertex pair is a syntactically correct and if that pair has the possibility of leading to a complete isomorphic mapping
class vf2_target_sets handles 1 and 2
class vf2_vertex_id_mapper handles 3
class vf2_isomorphism_feasibility_checker handles 4
The struct vf2_information holds objects of classes vf2_target_sets and vf2_vertex_id_mapper. 
The main function check_isomorphism uses this struct on its recursive calls and will pass along an object for class vf2_isomorphism_feasibility_checker. The code right now will not pass all the automated checks, but I need to know your thoughts on how I structured my code, the rest is an easy refactor. Thanks @joweich.

### CODE REVIEW

**Review by joweich — COMMENTED** [review]
Hey @sracha4355, very great to see that you built on the feedback from last time (https://github.com/bobluppes/graaf/pull/172#pullrequestreview-1728170848)! :rocket: 
I added a few more comments that are mostly easy fixes. However, I haven't had an in-depth look into the details of `graph_isomorphism_utils.tpp`. It's still a bit hard to follow what happens in the 500+ lines of code in this file. Given that this algorithm is quite complex, I feel you don't need to optimize/refactor every detail in this PR.  Just stop refactoring once you feel you're at a dead end :slightly_smiling_face: We can then do more optimizations in a follow-up PR.
To finally get you your well-deserved contribution to the main branch, I'd suggest to have a look into my latest feedback, add some more tests for minimal and more complex cases, and remove the debugging statements and commented-out code.
At some point, we'll also need a documentation page for the VF2 algorithm, but this can be done in a follow-up PR as well :slightly_smiling_face:

File: include/graaflib/algorithm/graph_isomorphism_util.tpp [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+		}
+	} else {
+		vertex_id_t original_vertex_id = graph2_reverse_mapping.at(vertex);
+		for(const auto& pair: graph2 -> get_vertices()){
+			if(pair.first != original_vertex_id && graph2 -> has_edge(pair.first, original_vertex_id)){
+				predecessors.insert(graph2_mapping.at(pair.first));
+			}
+		}
+	}
+	return std::move(predecessors);
```
[the remark this answers is not in the document]
Comment (sracha4355) [reply include/graaflib/algorithm/graph_isomorphism_util.tpp:53]: won't this change make it so our program returns a copy of predecessors?

File: include/graaflib/algorithm/graph_isomorphism.tpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	}
+	for (const auto& vertex : rhs.get_vertices()) {
+	std::size_t number_of_conn = rhs.get_neighbors(vertex.first).size();
+	rhs_node_degrees.push_back(number_of_conn);
+	}
+	std::sort(lhs_node_degrees.begin(), lhs_node_degrees.end());
+	std::sort(rhs_node_degrees.begin(), rhs_node_degrees.end());
+	for (int i = 0; i < lhs_node_degrees.size(); i++)
+	if (lhs_node_degrees[i] != rhs_node_degrees[i]) return false;
+	return true;
```
[the remark this answers is not in the document]
Comment (sracha4355) [reply include/graaflib/algorithm/graph_isomorphism.tpp:22]: this will tell the complier to go through each element in the vectors? thats cool.

File: include/graaflib/algorithm/graph_isomorphism_util.h [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+// Delete later
+#include <iostream>
+
+/**
+HELPER CLASSES
+*/
+
+namespace graaf::algorithm::vf2{
+
+struct predecessors_and_successors_of_vertex {
```
[the remark this answers is not in the document]
Comment (sracha4355) [reply include/graaflib/algorithm/graph_isomorphism_util.h:18]: I understand it is verbose, but for someone who has an idea of how the algorithm works, this name is pretty descriptive and says exactly what the struct is for. I would like to keep this as the name, if you would allow it.

**Comment by bobluppes** [comment]
Hi @sracha4355, sorry for the late reply and thanks a lot for working on this.

The implementation of the VF2 algorithm is definitely a lot more complex than I initially thought 😅 Nonetheless, very nice to see that you managed to implement this from scratch! Before merging this though, I would like to gain some more confidence in the correctness of the implementation by enhancing the unit tests.

Going forward, I would propose to close the current PR and to keep the branch as a reference implementation. Please feel free to ping me / re-open though if you would like to continue on this.

**Comment by sracha4355** [comment]
@bobluppes yeah the algorithm is definitely a beast of its own. I will let you know if I am available to work on it!
